File: cpp/238.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
  vector<int> productExceptSelf(vector<int>& nums) {
    int n = nums.size();
    vector<int> left{1}, right{1};
    for(int i=0; i<n; i++){
      left.push_back(nums[i] * left.back());
      right.push_back(nums[n-i-1] * right.back());
    }
    vector<int> res;
    for(int i=0; i<n; i++)
      res.push_back(left[i] * right[n-i-1]);
    return res;
  }
};

File: python/238.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        left = [1]
        n = len(nums)
        for i in range(n):
            left.append(nums[i]*left[-1])
        right = [1]
        for i in range(n-1, -1, -1):
            right.append(nums[i]*right[-1])
        res = []
        for i in range(n):
            res.append(left[i] * right[n-i-1])
        return res

print(Solution().productExceptSelf([1,2,3,4]))
